hkmc2/shared/src/main/scala/hkmc2/syntax/Tree.scala

```diff
@@ -239,35 +244,97 @@ enum Tree extends AutoLocated:
+      // understood as type names for unnamed contextual parameters:
+      // fun f(using A)
+      case ty: Ident if inUsing =>
+        R(TreeParam(flags, Ident(""), S(ty), N, modifiers))
+      // fun f(using A[B])
+      case ty @ TyApp(_, _) if inUsing =>
+        R(TreeParam(flags, Ident(""), S(ty), N, modifiers))
+
+      // fun f(a)
+      case id: Ident =>
+        R(TreeParam(flags, id, N, N, modifiers))
+      // fun f(a: A)
+      case InfixApp(id: Ident, Keyword.`:`, sign) =>
+        R(TreeParam(flags, id, S(sign), N, modifiers))
+
+      // fun f(..a)
+      case Spread(Keyword.`..`, _, S(id: Ident)) =>
+        R(TreeParam(flags, id, N, S(SpreadKind.Lazy), modifiers))
+      // fun f(...a)
+      case Spread(Keyword.`...`, _, S(id: Ident)) =>
+        R(TreeParam(flags, id, N, S(SpreadKind.Eager), modifiers))
+      // fun f(.._)
+      case Spread(Keyword.`..`, _, S(und: Under)) =>
+        R(TreeParam(flags, new Ident("_").withLocOf(und), N, S(SpreadKind.Lazy), modifiers))
+      // fun f(..._)
+      case Spread(Keyword.`...`, _, S(und: Under)) => 
+        R(TreeParam(flags, new Ident("_").withLocOf(und), N, S(SpreadKind.Eager), modifiers))
+      // fun f(..)
+      case Spread(Keyword.`..`, kwLoc, N) =>
+        R(TreeParam(flags, new Ident("_").withLoc(kwLoc), N, S(SpreadKind.Lazy), modifiers))
+      // fun f(...)
+      case Spread(Keyword.`...`, kwLoc, N) =>
+        R(TreeParam(flags, new Ident("_").withLoc(kwLoc), N, S(SpreadKind.Eager), modifiers))
+      
+      // * Unwrapping Cases
+      
+      // fun f(module <...>)
+      case TypeDef(Mod, inner, N) =>
+        go(inner, flags, modifiers + Mod)
+      // fun f(pattern <...>)
+      case TypeDef(Pat, inner, N) =>
+        go(inner, flags.copy(pat = true), modifiers + Pat)
+      // class C(val <...>)
+      case TermDef(ImmutVal, inner, _) =>
+        go(inner, flags.copy(isVal = true), modifiers + ImmutVal)
+      // class C(mut val <...>)
+      case TermDef(MutVal, inner, _) =>
+        go(inner, flags.copy(isVal = true, mut = true), modifiers + MutVal)
+      // fun f(using <...>)
+      case TermDef(Ins, inner, N) =>
+        go(inner, flags, modifiers + Ins)
+
+      // * Default Case
+      case _ => L:
+        ErrorReport:
+          msg"Expected a valid parameter, found ${this.describe}" -> this.toLoc :: Nil
+    
+    go(this, flags = FldFlags.empty.copy(isVal = inDataClass), modifiers = Set.empty)
+
+  def isModified(modifier: Keyword | DeclKind): Bool = this match
+    case td @ Tree.TypeDef(m, head, N) =>
+      (td.extension.isEmpty && td.withPart.isEmpty && m == modifier) || head.isModified(modifier)
+    case td @ Tree.TermDef(m, head, N) =>
+      (td.extension.isEmpty && td.withPart.isEmpty && m == modifier) || head.isModified(modifier)
+    case Modified(m, _, body) =>
+      modifier == m || body.isModified(modifier)
+    case _ =>
+      false
 
 object Tree:
+  // A parameter yet to be elaborated.
+  case class TreeParam(
```

Comment:
It should be called `ParamTree`, not `TreeParam`.

But actually, since it's already in `Tree`, why not just call it `Param`?

Reply:
> I think better name it `TreeParam` and move it outside the Tree, because not only in the elaborator there is `import Tree.*` but also some other places.

I made this change. Let me know if you like it